keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,365 @@
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
+* Some on node system workloads interested in pods readiness want to reduce resource consumption.
+
+## Proposal
+
+We are proposing to create a new kubelet API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
+* The API will be read-only and accessible for on-node workloads (we will use unix socket to connect) without the authorization.
+* The API will be versioned.
+
+### What kind of API to chose?
+
+Our goal is to expose the API locally on the node and without Control Plane dependency. That mean that we want to avoid authorization which will be done via kube-API. To fullfil this requirement we decided to use unix socket to access the API. On top of unix socket we can create REST or gRPC API.
+
+|                            |    gRPC | REST |
+|--------------------------- | --------- | ------|
+| Support for unix socket | yes | yes |
+| Can reuse existing API | possibly podresource API. See [here](#can-we-integrate-with-podresource-api) | no, the existing REST API is served on ReadOnlyPort (shutdown on various cloud providers due to security reasons), AuthenticatedPort (authorization is done by kube-API). |
+| Can reuse existing libraries (client-go, helpers in documentation) | no | yes|
+| Support for streaming | yes | yes|
+| strongly typed | yes | no |
+| versions support | yes | no |
+
+### Can we integrate with podresource API?
+
+|           | pod resource API | pod info API |
+| ----------| ---------------- | ------------ |
+| What information is returned with API | Container resources (devices, cpu_ids, memory, dynamic resources) | for Static and Regular Pods pod readiness Conditions and UID  |
+| Who is the owner of the data returned by API | kubelet, those data are not reported to Control Plane | static pods - kubelet, regular pods - kubelet, data are reported to kube-API |
+| What is the alternative for the API | reading the state files resource manager use | for regular pods - fetching pod's data with kube-API, for static pods - none |
+| Control Plane dependency | no | no |
+| Authorization | none, using unix socket to limit the clients to local workloads | none, using unix socket to limit the clients to local workloads |
+| access | Read-Only | Read-Only|
+
+Podresource API looks like a good fit for this endpoint. Our major concern is how will this integration impact the shape of this API in the future.
```

Comment:
I think I'd also note that we expect there to be little-to-no overlap in the clients for both of these APIs, they're both serving information that would interest different sets of clients. Admittedly this may converge over time as new features are requested, but it may make sense to keep these separate so clients don't need to fetch data they don't need.

Reply:
done